feat: ImportKeys command does not support updates

naveenpaul1 opened the pull request:
## Related issue

https://github.com/ory/hydra/issues/2436

@aeneasr  @Benehiko 

## Proposed changes

While importing keys keys are not update but removed if they were already presented.

Steps to reproduce the behavior:

Run keys import setName priv.pem pub.pem --skip-tls-verify once
Check setName is presented in table hydra_jwk
Run again keys import setName priv.pem pub.pem --skip-tls-verify
Check setName is not presented in table hydra_jwk

## Checklist

<!--
Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of
them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.
-->

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [x] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security. vulnerability, I
      confirm that I got green light (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [x] I have added tests that prove my fix is effective or that my feature
      works.
- [ ] I have added or changed [the documentation](docs/docs).

## Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution
you did and what alternatives you considered, etc...
-->

I think the issue is in cmd/cli/handler_jwk.go

var set jose.JSONWebKeySet
err = json.NewDecoder(response.Body).Decode(&set)
There as the keys were previously removed them response.Body contains the set key which the will be added again... So when set contains every key duplicated. I think this could be fixed by replacing:

var set jose.JSONWebKeySet
err = json.NewDecoder(response.Body).Decode(&set)
for:

set := jose.JSONWebKeySet{}

Commit:
feat: flush refresh tokens for service oauth2/flush

Commit:
feat: flush refresh tokens test added

aeneasr committed:
Merge branch 'master' into master

Commit:
fix : remove public prefix from kid

Commit:
Merge remote-tracking branch 'upstream/master'

# Conflicts:
#	oauth2/handler_test.go
#	persistence/sql/persister_oauth2.go

Commit:
fix : encode response body

Commit:
fix : ImportKeys command does not support updates

Commit:
fix : revert changes

Commit:
fix : ImportKeys command does not support updates

Commit:
fix : remove unwanted changes

aeneasr commented:
Thank you! Could you please add a failing test case which shows the broken behaviour so it is easier to understand what is going on and also prevent future regressions? :)

aeneasr committed:
Merge branch 'master' into IMPORTKEYS_UPDATES

Commit:
Merge remote-tracking branch 'upstream/master' into IMPORTKEYS_UPDATES
# Please enter a commit message to explain why this merge is necessary,
# especially if it merges an updated upstream into a topic branch.
#
# Lines starting with '#' will be ignored, and an empty message aborts
# the commit.

aeneasr committed:
Merge branch 'master' into IMPORTKEYS_UPDATES

Commit:
feat : add unit test

aeneasr commented on cmd/cli/handler_jwk_test.go:
```diff
@@ -60,4 +68,35 @@ func Test_toSDKFriendlyJSONWebKey(t *testing.T) {
+		NewHandler().Keys.ImportKeys(&cmd, []string{"setName", "../test/private_key.json", "../test/public_key.json"})
+		//running again to make sure the row in storage is not deleted
+		NewHandler().Keys.ImportKeys(&cmd, []string{"setName", "../test/private_key.json", "../test/public_key.json"})
+
```
We still need to test if the keys are in the store as expected right? :)

naveenpaul1 replied:
@aeneasr : unit test is updated, now we are checking the keys in db after key import

Commit:
feat : update unit test

aeneasr reviewed (changes requested):
I reverted the changes in handler_jwk to see if the test caused a failure and received the following error:

```
{
	"error": "Unable to insert or update resource because a resource with that value exists already",
	"error_description": ""
}
```

indicating that the key was already imported. The test does thus not describe the original bug which was reported as:

1. Run `hydra keys import setName priv.pem pub.pem --skip-tls-verify`
2. Run `hydra keys import setName priv.pem pub.pem --skip-tls-verify` again
3. JWK from the first / second call is not in the store

Instead I observed (in the test):

1. Run `hydra keys import setName priv.pem pub.pem --skip-tls-verify`
2. Run `hydra keys import setName priv.pem pub.pem --skip-tls-verify` again but fails because key already exists
3. JWK is still in the store

This also makes sense when looking at the test. The import is not a replace operation but instead appends the key(s) to the existing set - thus "echoing" the existing keys back to the API.

Could you maybe elaborate more specifically what is happening here?

naveenpaul1 commented:
> I reverted the changes in handler_jwk to see if the test caused a failure and received the following error:
> 
> ```
> {
> 	"error": "Unable to insert or update resource because a resource with that value exists already",
> 	"error_description": ""
> }
> ```
> 
> indicating that the key was already imported. The test does thus not describe the original bug which was reported as:
> 
> 1. Run `hydra keys import setName priv.pem pub.pem --skip-tls-verify`
> 2. Run `hydra keys import setName priv.pem pub.pem --skip-tls-verify` again
> 3. JWK from the first / second call is not in the store
> 
> Instead I observed (in the test):
> 
> 1. Run `hydra keys import setName priv.pem pub.pem --skip-tls-verify`
> 2. Run `hydra keys import setName priv.pem pub.pem --skip-tls-verify` again but fails because key already exists
> 3. JWK is still in the store
> 
> This also makes sense when looking at the test. The import is not a replace operation but instead appends the key(s) to the existing set - thus "echoing" the existing keys back to the API.
> 
> Could you maybe elaborate more specifically what is happening here?

naveenpaul1 commented:
> I reverted the changes in handler_jwk to see if the test caused a failure and received the following error:
> 
> ```
> {
> 	"error": "Unable to insert or update resource because a resource with that value exists already",
> 	"error_description": ""
> }
> ```
> 
> indicating that the key was already imported. The test does thus not describe the original bug which was reported as:
> 
> 1. Run `hydra keys import setName priv.pem pub.pem --skip-tls-verify`
> 2. Run `hydra keys import setName priv.pem pub.pem --skip-tls-verify` again
> 3. JWK from the first / second call is not in the store
> 
> Instead I observed (in the test):
> 
> 1. Run `hydra keys import setName priv.pem pub.pem --skip-tls-verify`
> 2. Run `hydra keys import setName priv.pem pub.pem --skip-tls-verify` again but fails because key already exists
> 3. JWK is still in the store
> 
> This also makes sense when looking at the test. The import is not a replace operation but instead appends the key(s) to the existing set - thus "echoing" the existing keys back to the API.
> 
> Could you maybe elaborate more specifically what is happening here?


> I reverted the changes in handler_jwk to see if the test caused a failure and received the following error:
> 
> ```
> {
> 	"error": "Unable to insert or update resource because a resource with that value exists already",
> 	"error_description": ""
> }
> ```
> 
> indicating that the key was already imported. The test does thus not describe the original bug which was reported as:
> 
> 1. Run `hydra keys import setName priv.pem pub.pem --skip-tls-verify`
> 2. Run `hydra keys import setName priv.pem pub.pem --skip-tls-verify` again
> 3. JWK from the first / second call is not in the store
> 
> Instead I observed (in the test):
> 
> 1. Run `hydra keys import setName priv.pem pub.pem --skip-tls-verify`
> 2. Run `hydra keys import setName priv.pem pub.pem --skip-tls-verify` again but fails because key already exists
> 3. JWK is still in the store
> 
> This also makes sense when looking at the test. The import is not a replace operation but instead appends the key(s) to the existing set - thus "echoing" the existing keys back to the API.
> 
> Could you maybe elaborate more specifically what is happening here?

@aeneasr : issue is in method `jwk.UpdateKeySet()`, where `DeleteKeySet()` is executing with value we are passing in `set` but `AddKeySet()` is failing because we are trying to add two key set with same `set` and `key` values and we are getting error you are mentioned in previous chat, at the same time already existing row is getting removed from db because of delete.
I tried to reproduce this in unit test but couldn't run the 
`m.GetKeySet(context.TODO(), "import-1")`
when the second import fails test case execution is getting stoped, since `cli,ImportKeys()` method exit the execution when update key set (PUT /keys/setName) retun not 200 in below line
`cmdx.CheckResponse(err, http.StatusOK, response)`,  Please correct me if am wrong

aeneasr commented:
Could you maybe push the test case you have problems with? That would help me understand what is failing! Thanks :)

Commit:
Merge remote-tracking branch 'upstream/master' into IMPORTKEYS_UPDATES

Commit:
feat : update unit test to for multiple key importing

Commit:
feat : update unit test to for multiple key importing

naveenpaul1 commented:
@aeneasr : I just updated the unit test case which covers multiple imports with the same key values, Since `ImportKeys()` method exit with `os.Exit(1)`, we need to catch it and execute the `GetKeySet()` method within the catch block to make sure imported key is not in db.

Commit:
Merge remote-tracking branch 'upstream/master' into IMPORTKEYS_UPDATES

fjvierap commented:
> Thank you! Could you please add a failing test case which shows the broken behaviour so it is easier to understand what is going on and also prevent future regressions? :)

Hi @aeneasr I was checking this and it is not easy to add a error test for this...  Note the code which was there:
```
func (h *JWKHandler) ImportKeys(cmd *cobra.Command, args []string) {
	cmdx.MinArgs(cmd, args, 2)

	id := args[0]
	use := flagx.MustGetString(cmd, "use")
	client := &http.Client{}

	/* #nosec G402 - we want to support dev environments, hence tls trickery */
	client.Transport = &http.Transport{
		TLSClientConfig: &tls.Config{
			InsecureSkipVerify: flagx.MustGetBool(cmd, "skip-tls-verify"),
		},
	}

	u := Remote(cmd) + "/keys/" + id
	request, err := http.NewRequest("GET", u, nil)
	cmdx.Must(err, "Unable to initialize HTTP request: %s", err)

	if flagx.MustGetBool(cmd, "fake-tls-termination") {
		request.Header.Set("X-Forwarded-Proto", "https")
	}

	if token := flagx.MustGetString(cmd, "access-token"); token != "" {
		request.Header.Set("Authorization", "Bearer "+token)
	}

	response, err := client.Do(request)
	cmdx.Must(err, "Unable to fetch data from %s: %s", u, err)
	defer response.Body.Close()

	if response.StatusCode != http.StatusOK && response.StatusCode != http.StatusNotFound {
		cmdx.Fatalf("Expected status code 200 or 404 but got %d while fetching data from %s", response.StatusCode, u)
	}

	set := jose.JSONWebKeySet{}
	for _, path := range args[1:] {
		file, err := ioutil.ReadFile(path)
		cmdx.Must(err, "Unable to read file %s", path)

		if key, privateErr := josex.LoadPrivateKey(file); privateErr != nil {
			key, publicErr := josex.LoadPublicKey(file)
			cmdx.Must(publicErr, `Unable to read key from file %s. Decoding file to private key failed with reason "%s" and decoding it to public key failed with reason: %s`, path, privateErr, publicErr)

			set.Keys = append(set.Keys, toSDKFriendlyJSONWebKey(key, "public:"+uuid.New(), use))
		} else {
			set.Keys = append(set.Keys, toSDKFriendlyJSONWebKey(key, "private:"+uuid.New(), use))
		}

		fmt.Printf("Successfully loaded key from file: %s\n", path)
	}

	body, err := json.Marshal(&set)
	cmdx.Must(err, "Unable to encode JSON Web Keys to JSON: %s", err)

	request, err = http.NewRequest("PUT", u, bytes.NewReader(body))
	cmdx.Must(err, "Unable to initialize HTTP request: %s", err)

	if flagx.MustGetBool(cmd, "fake-tls-termination") {
		request.Header.Set("X-Forwarded-Proto", "https")
	}

	if token := flagx.MustGetString(cmd, "access-token"); token != "" {
		request.Header.Set("Authorization", "Bearer "+token)
	}
	request.Header.Set("Content-Type", "application/json")

	response, err = client.Do(request)
	cmdx.CheckResponse(err, http.StatusOK, response)
	defer response.Body.Close()

	fmt.Println("JSON Web Key Set successfully imported!")
}
```
There for all error scenarios `cmdx.Must` is exiting the program with `cmdx.Must`How do you unit test those cases?

I checked already `root_test.go` which actually has this code:
```
err := rootCmd.Execute()
				if c.expectErr {
					assert.Error(t, err)
				} else {
					assert.NoError(t, err)
				}
```
Which I tested and `c.expectErr`is not used because failed scenarios exit the codes... and I think that is the reason there is no errors scenarios defined there.

fjvierap committed:
Adjust unit test

aeneasr commented:
Hm yeah the problem is that I think that this PR introduces other issues with the import command because it overwrites a key completely? And what we probably want is to append the imported keys to the list. I also checked the HTTP handlers. The method used is

https://github.com/ory/hydra/blob/64a332a98fc1c3a73e4b39b58b21b4cd61f5b240/jwk/handler.go#L246-L266

Here the handler is removing the key set and then adding all the imported keys. Therefore, the current behaviour is correct in my opinion.

I think the problem described in the opening PR is that, if you import a key that already exists, that fails, because the key is now being sent as a duplicate. The proper resolution for that should be to check in the import handler if the imported keys match one of the existing keys, and update those.

Therefore, we definitely need a test to cover this. The easiest method would probably be to have `ImportKeys(cmd *cobra.Command, args []string) error` instead and not rely on `cmd.Must()`. This would be a bit of refactoring though.

I'll mark this PR as a draft for now because it will introduce new bugs (removing existing keys) and the bug this PR solves is a special case that should be handled differently!

aeneasr commented on cmd/cli/handler_jwk.go:
```diff
@@ -123,10 +123,7 @@ func (h *JWKHandler) ImportKeys(cmd *cobra.Command, args []string) {
-	err = json.NewDecoder(response.Body).Decode(&set)
-	cmdx.Must(err, "Unable to decode payload to JSON: %s", err)
-
+	set := jose.JSONWebKeySet{}
 	for _, path := range args[1:] {
```
Instead of ignoring the existing keys we should make sure that we update any keys that have a matching key id :)

TimoDoernenburg replied:
Hey, I updated the branch and added an update function in the importer. Existing keys (same KeyID and SID) will be replaced and new keys (new KeyID and same SID) will be added to the list.
I also updated the test to test the changes.

TimoDoernenburg committed:
Merge branch 'master' into IMPORTKEYS_UPDATES

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ory/hydra?pullRequest=2460) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/ory/hydra?pullRequest=2460) before we can accept your contribution.<br/>**3** out of **4** committers have signed the CLA.<br/><br/>:white_check_mark: fjvierap<br/>:white_check_mark: aeneasr<br/>:white_check_mark: TimoDoernenburg<br/>:x: naveenpaul1<br/><hr/>**naveenpaul1** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ory/hydra?pullRequest=2460) it.</sub>

Commit:
Merge branch 'master' of https://github.com/ory/hydra into IMPORTKEYS_UPDATES

Commit:
updating existing keys during import instead of replacing the list

Commit:
Merge branch 'IMPORTKEYS_UPDATES' of https://github.com/naveenpaul1/hydra into IMPORTKEYS_UPDATES

TimoDoernenburg committed:
Merge branch 'master' into IMPORTKEYS_UPDATES

aeneasr committed:
Merge branch 'master' into IMPORTKEYS_UPDATES

aeneasr commented:
@TimoDoernenburg - thank you! I merged with master, let's see if the CI passes :) The CLA also has an issue:

Thank you, this looks great! It looks like the CLA bot is not properly detecting your signature. To fix this, try the following:

```
$ git commit  --amend --author="Author Name <email@address.com>"
```

Ensure that `Author Name` is replaced with your GitHub username (e.g. `aeneasr`) and that the email address is replaced with the email address you have set up in GitHub (e.g. `3372410+aeneasr@users.noreply.github.com`):

```
$ git commit  --amend --author="aeneasr <3372410+aeneasr@users.noreply.github.com>"
```

Once that is done, you can force-push your changes (make sure to push to the correct remote and branch!):

```
$ git push --force <remote> HEAD:<branch>
```

aeneasr commented on cmd/cli/handler_jwk.go:
```diff
@@ -89,6 +89,21 @@ func toSDKFriendlyJSONWebKey(key interface{}, kid, use string) jose.JSONWebKey {
 	}
 }
 
+func updateKey(set jose.JSONWebKeySet,newKey jose.JSONWebKey) []jose.JSONWebKey {
```
It would be great if you could add a unit test for this function :)

aeneasr reviewed (changes requested):
Thank you @TimoDoernenburg - this looks much better now! :) Got some more ideas how to improve it!

TimoDoernenburg committed:
Merge branch 'IMPORTKEYS_UPDATES' of https://github.com/naveenpaul1/hydra into IMPORTKEYS_UPDATES

Commit:
Merge branch 'master' of https://github.com/ory/hydra into IMPORTKEYS_UPDATES

TimoDoernenburg committed:
add updateKey test

TimoDoernenburg commented on cmd/cli/handler_import_jwk_test.go:
```diff
@@ -52,3 +53,40 @@ func Test_toImportJSONWebKey(t *testing.T) {
 
 }
+
+func Test_toUpdateKeyList(t *testing.T) {
```
I added another test for the importer updatekey function.

aeneasr committed:
Merge branch 'master' into IMPORTKEYS_UPDATES

aeneasr commented:
Thank you! So is this ready for another review? Sorry that I missed this, since this is not your PR I assume you can't mark this as "Ready for Review". PRs that are marked as draft are not shown on my dashboard :/

TimoDoernenburg committed:
Merge branch 'master' of https://github.com/ory/hydra into IMPORTKEYS_UPDATES

aeneasr commented:
Thank you, could you please address the failing test? 🙌 

https://app.circleci.com/pipelines/github/ory/hydra/2776/workflows/f718b951-bbac-4f1d-9839-37847c217bc9/jobs/32588

aeneasr commented on cmd/cli/handler_import_jwk_test.go:
```diff
@@ -0,0 +1,92 @@
+
+}
+
+func Test_toUpdateKeyList(t *testing.T) {
```
```suggestion
func TestUpdateKeyList(t *testing.T) {
```

aeneasr commented on cmd/cli/handler_import_jwk_test.go:
```diff
@@ -0,0 +1,92 @@
+}
+
+func Test_toUpdateKeyList(t *testing.T) {
+
```
```suggestion
```

aeneasr commented on cmd/cli/handler_import_jwk_test.go:
```diff
@@ -0,0 +1,92 @@
+	set.Keys = updateKey(set,toSDKFriendlyJSONWebKey(anotherPEM, "different-sid", "sig") )
+	assert.Equal(t, len(set.Keys),3)
+
+
```
Add a check that `another-sid` is still `dummyPEM` :)

aeneasr commented on cmd/cli/handler_jwk.go:
```diff
@@ -89,6 +89,21 @@ func toSDKFriendlyJSONWebKey(key interface{}, kid, use string) jose.JSONWebKey {
 	}
 }
 
+func updateKey(set jose.JSONWebKeySet, newKey jose.JSONWebKey) []jose.JSONWebKey {
```
```suggestion
// updateKey replaces a key in the JSON Key Set or appends it if the key ID is not yet known.
func updateKey(set jose.JSONWebKeySet, newKey jose.JSONWebKey) []jose.JSONWebKey {
```

aeneasr reviewed (changes requested):
Only some minor code format things (run `make format` to auto-format the code). Also one, last test - then I will merge this! :)

TimoDoernenburg committed:
Update cmd/cli/handler_jwk.go

Co-authored-by: hackerman <3372410+aeneasr@users.noreply.github.com>

TimoDoernenburg committed:
Update cmd/cli/handler_import_jwk_test.go

Co-authored-by: hackerman <3372410+aeneasr@users.noreply.github.com>

TimoDoernenburg committed:
Update cmd/cli/handler_import_jwk_test.go

Co-authored-by: hackerman <3372410+aeneasr@users.noreply.github.com>

TimoDoernenburg committed:
Update cmd/cli/handler_import_jwk_test.go

Co-authored-by: hackerman <3372410+aeneasr@users.noreply.github.com>

TimoDoernenburg committed:
Update cmd/cli/handler_import_jwk_test.go

Co-authored-by: hackerman <3372410+aeneasr@users.noreply.github.com>

aeneasr commented:
Thank you, this looks great! The CI is failing because some files are formatted incorrectly. To format them, run:

```
$ make format
$ git commit -a -m "styles: format code"
$ git push
```

Thank you!

TimoDoernenburg committed:
expand test, fix goimports-ed error

TimoDoernenburg committed:
expand test, fix goimports-ed error

TimoDoernenburg commented:
The tests are passing now.

aeneasr committed:
Merge branch 'master' into IMPORTKEYS_UPDATES

aeneasr committed:
chore: code review

aeneasr reviewed (approved):
Thank you all!
